### ISSUE TITLE
🐛 Add support for Android in the compiler (termux)

### ISSUE THREAD

**Issue by sebastianjnuwu** [rust-lang/rust#119306]
I was observing some issues from other people and it always said that the compiler was not supported as mentioned in the documentation, is there any alternative? 

![Screenshot_20231225-173358~2](https://github.com/rust-lang/rust/assets/101356556/234331b0-8fb1-4646-82fb-80375becd039)

```[tasklist]
### Tasks
```

**Comment by sebastianjnuwu** [comment]
You will probably deny it, but it is extremely important to add a compiler for Android as you can see there are excellent projects that use rust, but rust ends up becoming a thorn in the side. Implement compiler for Android add support for this for mobile developers!

**Comment by saethlin** [comment]
> You will probably deny it

No need to start by assuming bad faith here.

We already support multiple Android targets, they're just tier 2 not tier 2 with host tools. You can read about the requirements here: https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html#tier-2-with-host-tools

**Comment by Nilstrieb** [comment]
The way target things usually work is that someone that needs it makes a PR raising the tier of the target with justification for every bullet point in the target tier policy that entails. The compiler team does not have the capacity to add and raise targets for others, this effort usually has to come from people who want it (like you :)).

**Comment by sebastianjnuwu** [comment]
> The way target things usually work is that someone that needs it makes a PR raising the tier of the target with justification for every bullet point in the target tier policy that entails. The compiler team does not have the capacity to add and raise targets for others, this effort usually has to come from people who want it (like you :)).

Not just anyone can do this! For example, I don't have the necessary knowledge of how to do this, it works alone, I asked a friend for help to try to do it but he said it wasn't random that do this

**Comment by bjorn3** [comment]
Termux itself already builds rustc for android. You can use `pkg install rust` in termux to install it. This doesn't help you with beta and nightly releases, but at least they ship the latest stable.

**Comment by sebastianjnuwu** [comment]
> Termux itself already builds rustc for android. You can use `pkg install rust` in termux to install it. This doesn't help you with beta and nightly releases, but at least they ship the latest stable.

```
~/biome-main $ cargo   
info: syncing channel updates for '1.74.0-armv7-linux-androideabi'
info: latest update on 2023-11-16, rust version 1.74.0 (79e9716c9 2023-11-13)                                                 
error: target 'armv7-linux-androideabi' not found in channel. 
```

**Comment by sebastianjnuwu** [comment]
https://github.com/biomejs/biome/issues/1340

**Comment by bjorn3** [comment]
You need to uninstall rustup first for the termux packaged rust to be used.

**Comment by maurer** [comment]
tl;dr: We're not currently interested in making Android "Tier 2 with Host Tools". If we were, it would be `aarch64` not `armv7`, and I don't think the resources would be justified even if someone else provided it at the moment.

Summarizing a few points for @sebastianjnuwu who may not have been able to glean everything from context:

# Background
This stuff might not be immediately familiar to people who normally operate in the web sphere rather than the systems one.

## Platforms
Build processes for compilers classically referenced three platforms:
- `build` - the platform on which the compiler itself is being built
- `host` - the platform on which the compiler being built expects to run
- `target` - the platform for which the compiler being built expects to generate code

Modern compilers (especially LLVM-based ones like `rustc` or `clang`) tend to support multiple `target` platforms within the same binary, but otherwise still follow this pattern.

## Rust Tiered Support
Rust has [three tiers](https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html) of platform support. While the details are more complex, it basically boils down to:

1. CI runs tests *on the platform*.
2. CI runs builds on a supported platform (`x86_64-unknown-linux` in practice) of code for the target platform, performing any checks that can be done without running it.
3. Support for the platform is present in the codebase, but not tested.

# Your Initial Request
Initially, you seemed to want `rustup` to be able to install a host=`armv7-linux-androideabi` target=`armv7-linux-androideabi` compiler. `rustup` installs from a manifest that contains host builds for all "Tier 1 with Host Tools" and "Tier 2 with Host Tools". Since it's strictly harder to do the former than the latter, we'll assume you want the latter. All `android` targets are currently "Tier 2". Configuring the `host` to `armv7-linux-androideabi` for tools is believed to work, but is untested. Upgrading them to [Tier 2 with Host Tools](https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html#tier-2-with-host-tools) requires few major pieces:

- "The host tools must build and run reliably in CI" - the *run* portion is the critical piece here. If we wanted to make `armv7-linux-androideabi` a "Tier 2 with Host Tools" target, someone would need to provide the Rust project with a set of 32-bit ARM-supporting Android devices to test-run the built compiler. We don't currently have a plan to do this, and while I might be mistaken, it doesn't sound like you have capacity available for this either.
- "Building host tools for the target must not take substantially longer than building host tools for other targets" - this is especially hard for the host target you've chosen, as high-end Android devices are dropping support for 32-bit execution. Even if we were going to go through the overhead of setting up a fleet of Android devices, high-end android devices are moving to only support a 64-bit userspace. It might be possible to mitigate this by locating an `aarch64` cloud provider and run a 32-bit userspace on it via virtualization, but most cloud providers don't provide nested virtualization. This might be solvable, but it just makes the problem harder.
- "There must be a reasonable expectation that the host tools will be used, for purposes other than to prove that they can be used." - you are one such user, but most users don't develop *on* Android devices, they develop *for* them. This means their `host` would be set to a traditional Linux/Windows/Mac device, and `target` would be set to an Android-based target. If you have statistics indicating that building code on Android devices is popular, I'd love to see it.

# Running Biome on Android (What I think you actually want to do)

1. Run `which rustc` and `which cargo`.
2. If either return `$HOME/.cargo/bin/$TOOL`, run `rustup self uninstall`
3. Run `pkg install rust` as suggested above
4. Check that `rustc` can build a a sample file (put `fn main() {}` in a file called `foo.rs`, and run `rustc foo.rs`)
5. Use Biome / `npx` as normal.
6. If it still gives the error, Biome or one of its dependencies is attempting to force the use of `rustup` directly rather than respecting the package manager. In that case, you can either use the instructions for [custom toolchains](https://rust-lang.github.io/rustup/concepts/toolchains.html#custom-toolchains) to name a local toolchain the same way `Biome` expects to find one, or file a bug against whoever is trying to force `rustup` usage rather than allowing the user to provide one.

# Other possibilities
I do want to note that I think you'll have a better experience if you use your phone as an SSH terminal to a traditional Linux server for development rather than doing the actual development on your phone. If you don't have your own, most cloud providers will give you a small one for less than $10/mo. Especially if you're using a tool like `npx` which attempts to download and run tools for you automatically, you're not going to have a good time on an uncommon platform even after you get past the issue you're having configuring a Rust compiler.

If the goal is not to develop on your phone, but to *deploy* on your phone, you can do that without any of this. Build and bundle your website on PC, then push the built site to your device and just serve it there rather than compiling it there.

**Comment by RalfJung** [comment]
I'll close this then since we're unlikely to add host support for Android any time soon, and we already do support building for Android as a target. npx and biome might not support cross-compilation, but then that's a missing feature on their side; for mobile platforms, it is not reasonable to expect the compiler to run on the device itself.
